### DISCUSSION TITLE
Does bevy provide an API similar to unity's GrabPass?

### DISCUSSION THREAD

**Discussion by Ahuang0107** [bevyengine/bevy#16008] [Q&A] [answered]
I'm trying to achieve photoshop's layer blend modes effect with bevy sprite, like overlay, soft light, hard light, which I often use in unity.
In Unity, it mainly relies on the GrabPass command of the built-in pipeline, which will grab the screen once before each batch process that uses GrabPass. But i can not find same api or alternatives in bevy, is there anyway i can do this?
[Unity GrabPass Doc](https://docs.unity3d.com/Manual/SL-GrabPass.html)

this is the blend modes i want to achieve, mainly overlay, soft light, hard light.
<img width="307" alt="Screenshot 2024-10-20 021242" src="https://github.com/user-attachments/assets/b3e1b1fa-47c4-4552-bfe3-a899029ea3fe">
<img width="510" alt="Screenshot 2024-10-20 023317" src="https://github.com/user-attachments/assets/5f1e0f72-86c7-4974-a43f-cf3683f113e2">

this is the blend mode component i used in unity, which used GrabPass in his shader.
<img width="458" alt="Screenshot 2024-10-20 021051" src="https://github.com/user-attachments/assets/c974ac32-d09c-4d95-ace8-72f01a0d0f37">
<img width="583" alt="Screenshot 2024-10-20 021508" src="https://github.com/user-attachments/assets/478289c5-09e1-47e8-8baf-43542c942f1d">

**Comment by JMS55** [comment]
No there is not. It also generally performs poorly and should be avoided if possible in Unity.

**Reply by Ahuang0107** [reply]
i find a way to achieve it base on wgpu, but i don not know how to do in bevy. any advice for me? Details are as follows. my current simple idea is that if the mutable reference command_encoder belong to RenderContext  can be accessed in the render function of RenderCommand, this problem can be solved. Of course I'm just a game developer not an game engine developer. I don't know if this will cause other problems.

**Reply by JMS55** [reply, marked as answer]
You'll probably need to rip out the existing opaque render pass node, and add your own that splits the draws up into multiple different render passes, inserting texture copies in between.

There's nothing built into bevy for this.

**Comment by Ahuang0107** [comment]
hi there again, after a few days of exploration, I find a way to achieve blend mode by grab screen texture base on wgpu. but i don know how to do it in bevy, i try to modify SpritePipeline, but in the end I found that I could not get the command_encoder in any phases outside the bevy_render module to perform the encoder.copy_texture_to_texture operation. is there any advice?

this is the code base on wgpu.
<img width="539" alt="Screenshot 2024-10-24 174102" src="https://github.com/user-attachments/assets/816ee89d-d3fe-4ecc-882d-74c5628b08d1">

i try to create a transparent2d RenderCommand for grabing screen and copy it into a texture, but it turns out that 
the command_encoder.copy_texture_to_texture in the follow will queue in the render_queue first.
<img width="473" alt="Screenshot 2024-10-24 175017" src="https://github.com/user-attachments/assets/f39107cb-46ec-4c23-bbf5-885256fa9443">

**Comment by Ahuang0107** [comment]
emmm, i dont now why wgpu::RenderPass have a mutable reference of CommandEncoder. if i can access this command_encoder, the problem will solve. unfortunately it is belong to wgpu not bevy.
